domain/testing/watcher.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package testing
+
+import (
+	"fmt"
+
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/core/watcher/watchertest"
+)
+
+// NamespaceWatcherFactory implements a simple NamespaceWatcher creation process
+// for tests. It supports channels of changes on to multiple namespaces.
+// Channels created on this factory are buffered by 1 slot. This means that data
+// needs to be consumed from the watcher before the next lot of data is placed
+// to avoid a single goroutine lock.
+type NamespaceWatcherFactory struct {
+	InitialStateFunc func(string) ([]string, error)
+	WatcherChans     map[string]chan []string
+}
+
+// Close must be called after using this factory to safely clean up all the
+// channels.
+func (f *NamespaceWatcherFactory) Close() {
+	for _, ch := range f.WatcherChans {
+		close(ch)
+	}
+}
+
+// FeedChange feeds the supplied change for the given namespace into any
+// watchers that are open. If no watchers exist this func call evaluates to a
+// noop.
+func (f *NamespaceWatcherFactory) FeedChange(
+	namespace string,
+	_ changestream.ChangeType,
+	change []string) {
+	ch, exists := f.WatcherChans[namespace]
+	if !exists {
+		return
+	}
+	ch <- change
```

Comment:
no naked channel operations

Reply:
I've plumbed up the context like I should have done :)